tools/tpm2_stirrandom.c

```diff
@@ -0,0 +1,149 @@
+// 2. Redistributions in binary form must reproduce the above copyright notice,
+// this list of conditions and the following disclaimer in the documentation
+// and/or other materials provided with the distribution.
+//
+// 3. Neither the name of Intel Corporation nor the names of its contributors
+// may be used to endorse or promote products derived from this software without
+// specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
+// THE POSSIBILITY OF SUCH DAMAGE.
+//**********************************************************************;
+
+#include <ctype.h>
+#include <errno.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+
+#include <tss2/tss2_esys.h>
+
+#include "log.h"
+#include "files.h"
+#include "tpm2_options.h"
+#include "tpm2_tool.h"
+#include "tpm2_util.h"
+
+/* Spec enforce input data to be not longer than 128 bytes */
+#define MAX_SIZE_TO_READ 128
+
+typedef struct tpm_stirrandom_ctx tpm_stirrandom_ctx;
+struct tpm_stirrandom_ctx {
+    TPM2B_SENSITIVE_DATA    in_data;
+    char*                   input;
+    UINT16                  size_to_read;
+    struct {
+        UINT8 I : 1;
+        UINT8 s : 1;
+        UINT8 Unused : 6;
+    } flags;
+};
+
+static tpm_stirrandom_ctx ctx;
+
+static bool do_stirrandom(ESYS_CONTEXT *ectx) {
+    TSS2_RC rval = Esys_StirRandom(ectx, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &ctx.in_data);
+    if (rval != TSS2_RC_SUCCESS) {
+        LOG_ERR("Error while injecting specified additionnal data into TPM random pool");
+        LOG_PERR(Esys_StirRandom, rval);
+        return false;
+    }
+
+    return true;
+}
+
+static bool on_option(char key, char *value) {
+    bool result;
+
+    switch (key) {
+    case 'I':
+        ctx.input = value;
+        ctx.flags.I = 1;
+        break;
+    case 's':
+        result = tpm2_util_string_to_uint16(value, &ctx.size_to_read);
+        if (!result) {
+            LOG_ERR("Could not convert size to number, got: \"%s\"",
+                    value);
+            return false;
+        }
+        if(ctx.size_to_read > MAX_SIZE_TO_READ){
+            LOG_WARN("Requested size to read is too long. Set size to 128");
+            ctx.size_to_read = MAX_SIZE_TO_READ;
+        }
+        ctx.flags.s = 1;
```

Comment:
Both of these options should go away... the interface should be:

Use stdin (no argument via pipe)
```
dd if=/dev/urandom bs=1 count=32 | tpm_stirandom
```

Use stdin (no argument via file)
```
dd if=/dev/urandom bs=1 count=256 of=rand.dat
tpm_stirandom rand.dat < rand.dat
```

use a single file (an argument)
```
dd if=/dev/urandom bs=1 count=192 of=rand.dat
tpm_stirandom rand.dat
```

Fails because too big
```
dd if=/dev/urandom bs=1 count=257 of=rand.dat
tpm_stirandom rand.dat
```

BOTH calls should be bounded by the buffer size of a `TPM2B_SENSITIVE_DATA`  type.

Reply:
This is what i had in mind ! Thank you for highlighting :)